### PR TITLE
feat(cast): add ENS resolution for function parameters

### PR DESCRIPTION
**Pull request by ncitron**
## Motivation
ENS resolution is not supported for function parameters in `cast call` and `cast send`. To see this error in actions, run `cast call dai balanceOf ncitron.eth`.

## Solution
This problem is solved by preprocessing all arguments in `build_tx` to check if the parameter contains the string ".eth". If it does, we automatically attempt to resolve the name.

### CODE REVIEW

File: cast/src/lib.rs [hunk 1]
```diff
@@ -271,13 +272,28 @@ where
         };
 
         let func = if let Some((sig, args)) = args {
+            let args = join_all(args.iter().map(|arg| async {
+                if arg.ends_with(".eth") {
+                    let val = format!(
+                        "0x{}",
+                        hex::encode(self.provider.resolve_name(arg).await?.as_bytes())
```
Comment (mattsse) [inline cast/src/lib.rs:279]: can we add a verbose error message here, somehting like "Failed to resolve {}"?

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks, 1 suggestions

**Comment by mds1** [comment]
A few thoughts on this PR:

1. Should auto-resolution even be built in at this level? Currently you can already do `cast call dai "balanceOf(address)(uint256)" $(cast resolve-name ncitron.eth)` to get the same functionality, so perhaps we should get some thoughts around the cast design philosophy before merging? For example, this likely would never be merged in seth since they follow the unix philosophy, but this would be supported in ethers.js where ENS names can be used anywhere an address can be used. What's cast's philosophy here? Native ENS resolution is already supported in some places (e.g. we can do `cast call ethereum.eth "m_numOwners()(uint256)"` and it resolves ethereum.eth), so maybe the ethers.js approach is more in-line with the cast approach? Or maybe the `resolve_addr` method used should pulled into it's own cast command instead of auto-resolving? Perhaps worth a discussion since I think it's been a bit ad-hoc so far, and I think consistency is important both for UX and maintainability 

2. If we do want ENS names to valid anywhere, I'm not sure if this is the right way to implement. This adds support just to cast methods that pass through `build_tx` and only for `.eth.` names, but introduces inconsistency since addressbook values wouldn't be supported as `cast call` parameters, and ENS+addressbook wouldn't be supported in methods like `cast calldata` or `cast abi-encode`. If we go with the ethers.js "ENS as first class citizens" approach:
    1. Any domain is potentially a valid ENS name, so we shouldn't be only checking `.eth`. You could just pass any address type arg arg to `foundry_utils::resolve_addr` and this will automatically handle all domains + valid addressbook values
    2. Probably need a pass through all `cast` commands to ensure consistency, and to replace addresses with `foundry_utils::resolve_addr(addr, chain)` each time?

**Comment by gakonst** [comment]
On philosophy: I Personally prefer being more opinionated for the user and doing the resolution for them. It is harder to do that in Seth because scripts tend to get more complex when you try to handle multiple cases afaict, whereas here we have a full programming language on our hands - we should use it!

To the other point: let's refactor the name resolution to a helper function, have it work with dot eth initially, and expand the helper function over time with address book / other domains?

+1 on doing this in every place where an address can appear

**Comment by mds1** [comment]
`cast resolve-name vitalik.argent.xyz` successfully resolves, so we should be able to support non `.eth` domains as well for free. But otherwise that all sounds good to me—thanks!

**Comment by ncitron** [comment]
I wonder if it would be better to handle name resolution here https://github.com/gakonst/foundry/blob/master/utils/src/lib.rs#L376? It is a bit of a lower level function, so will hopefully encompass more commands in the future, and it ensures that we don't accidentally do ENS resolution for a string that contains ".eth". This was my main reservation for not resolving all ENS names, since it is possible that the argument was actually a string containing ".eth" rather than an address. With this method we can safely check if an address contains a "." without worrying about clashing with normal strings. What do you think?

**Comment by gakonst** [comment]
Yeah @ncitron if adding a provider object to the helper function and doing the resolution at the lowest level ppossible

**Review by gakonst — APPROVED** [review]

**Comment by gakonst** [comment]
tried it locally `cast call dai balanceOf vitalik.eth` is really slick

**Pull request merged**